Python/chapter02/2.6 - Palindrome/miguel_2.6_soln.py

```diff
@@ -154,6 +155,29 @@ def reverse_linked_list(ll: LinkedList) -> LinkedList:
     return output_ll
 
 
+def palindrome_constant_space(ll: LinkedList) -> bool:
+    """
+    Same as the palindrome function below, but
+    with constant space
+    Runtime:  O(n^2)
```

Comment:
is this optimal?

Reply:
yes

Reply:
can you justify your answer?

Reply:
The outer loop will be doing n/2 steps.  As we get closer to n/2, there will be less steps (steps = advancing n2 pointer).  The number of steps at each iteration is n - 2*i - 1.  The sequence will look like this:  (n-1) + (n-3) + (n-5) + ... + 5 + 3 + 1.  So, the number of steps will be the sum of that sequence.  We can add the same sequence to itself and then rearrange to get (n + n + n + ... + n + n + n), and since we have n/2 terms, we get n * n/2 = n^2/2.  But we need to divide by 2 again to undo the adding sequence to itself part, which results in n^2 / 4.  I could be wrong, but this is my reasoning.

Reply:
Oh wait...

Reply:
I answered the wrong question, you asked if this is optimal, not to justify my runtime analysis.

Reply:
that is a decent justification of why the current code is O(n^2), but not whether it's optimal!

i'll ask in a different way: can you think of an O(1) space/time way to compare the head and tail of the linked list for equality?

Reply:
O(1) space and O(1) time way?  Or do you mean O(1) but better than O(n^2) time way?

Reply:
Head and tail can be compared in constant time, but I'm not sure about the rest.  If this was a doubly linked list, I can get a O(n) runtime with O(1) space.

Reply:
the former.

Reply:
Well for the first and last elements we can access head and tail in constant time.  However, for the middle elements, I don't think the complement can be accessed in constant time unless we had a doubly linked list and 'popped' the head and tail each time.

Reply:
you already have a method that can reverse the linked list in-place. what if you only reversed half?

Reply:
right, I forget that creating a new LinkedList() doesn't always use n space.  Solution updated